src/platform/w3dengine/client/w3dtreebuffer.cpp

```diff
@@ -0,0 +1,1553 @@
+            speed = module->m_minimumToppleSpeed;
+        }
+
+        tree->topple_direction = *pos;
+        tree->topple_direction.Normalize();
+        tree->angular_accumulation = 0.0f;
+        tree->angular_velocity = speed * module->m_initialVelocityPercent;
+        tree->angular_acceleration = speed * module->m_initialAccelPercent;
+        tree->topple_state = TTree::TOPPLE_FALLING;
+        tree->options = options;
+        Coord3D primary;
+        primary.Set(tree->location.X, tree->location.Y, tree->location.Z);
+        Do_FX_Pos(module->m_toppleFX, &primary, nullptr, 0.0f, nullptr, 0.0f);
+        m_vertexesDirty = true;
+        tree->topple_transform.Make_Identity();
+        tree->topple_transform.Set_Translation(tree->location);
+    }
+}
+
+void W3DTreeBuffer::Update_Toppling_Tree(TTree *tree)
+{
+    captainslog_dbgassert(tree->topple_state != TTree::TOPPLE_UPRIGHT, "bad topple state");
+    if (tree->topple_state != TTree::TOPPLE_UPRIGHT && tree->topple_state != TTree::TOPPLE_UNK4) {
+        W3DTreeDrawModuleData *module = m_treeTypes[tree->tree_type].module;
+        int id;
+
+        if (g_thePlayerList != nullptr) {
+            id = g_thePlayerList->Get_Local_Player()->Get_Player_Index();
+        } else {
+            id = 0;
+        }
+
+        Coord3D loc;
+        loc.Set(tree->location.X, tree->location.Y, tree->location.Z);
+
+        if (g_thePartitionManager->Get_Shroud_Status_For_Player(id, &loc) == SHROUDED_UNK3) {
+            tree->topple_state = TTree::TOPPLE_UNK2;
+        } else if (tree->topple_state == TTree::TOPPLE_UNK2) {
+            tree->angular_velocity = 0.0f;
+            tree->topple_state = TTree::TOPPLE_UNK4;
+            tree->topple_transform.In_Place_Pre_Rotate_X(-1.521709f * tree->topple_direction.y);
+            tree->topple_transform.In_Place_Pre_Rotate_Y(1.521709f * tree->topple_direction.x);
+
+            if (module->m_killWhenFinishedToppling) {
+                tree->topple_sink_countdown = 0;
+            }
+        } else {
+            float angular_velocity = tree->angular_velocity;
+
+            if (angular_velocity + tree->angular_accumulation > 1.521709f) {
+                angular_velocity = 1.521709f - tree->angular_accumulation;
+            }
+
+            tree->topple_transform.In_Place_Pre_Rotate_X(-angular_velocity * tree->topple_direction.y);
+            tree->topple_transform.In_Place_Pre_Rotate_Y(angular_velocity * tree->topple_direction.x);
+            tree->angular_accumulation = angular_velocity + tree->angular_accumulation;
+
+            if (tree->angular_accumulation < 1.521709f || tree->angular_velocity <= 0.0f) {
+                tree->angular_velocity = tree->angular_velocity + tree->angular_acceleration;
+            } else {
+                tree->angular_velocity = -module->m_bounceVelocityPercent * tree->angular_velocity;
+
+                if ((tree->options & 1) != 0 || (0.0099999998f > GameMath::Fabs(tree->angular_velocity))) {
```

Comment:
flip, also something may be off with (tree->options & 1) != 0, original is ((tree->m_options & 1) != 0) == 1

Reply:
fixed.